### DISCUSSION TITLE
Different mapping if 'held' or 'alone' - Timeout modifier not working as intended?

### DISCUSSION THREAD

**Discussion by neophrema** [xremap/xremap#614] [Ideas]
Hello,
thank you very much for this program, i was looking for this for a long (!) time.

I am using QMK on my external keyboard and I have mapped "F" and "J" to be registered as Shift, if I hold these keys instead of just typing. (This is due to my pinkies hurting).
I found the same functionality in xremap, but I can't use it due to timing issues.

```
modmap:
  - name: ShiftJ 
    remap: 
      KEY_J:
        held: SHIFT_R
        alone: KEY_J
      # alone_timeout_millis: 10000 # Optional
```

I'm not sure if I have a feature request or a bug complaint.

If I'm fluently typing the word: "jelly"
It registers instead:                 "Elly"

1.) I think it should be specified how long the key has to be pressed, before it gets registered as held.
1.2) Is there a variable "held_timeout_millis"?

2.) I do not understand how "alone_timeout_millis" works.
2.1)I played with different values and it seems like xremap cap these values, correct?
2.2)Im not sure if the variable name - "(alone_)timeout(_millis)" -  might be ambiguous.

I think it would be ideal if I could set the time that the button needs to be pressed, e.g. 3 sec, before it acts as a modifier, before that its just an letter, e.g. "j".
I understand that I have a rare user case, yet I suspect that there *could* be an issue with the code.

Hopefully this might be interesting for others as well.
neophrema

**Comment by k0kubun** [comment]
> If I'm fluently typing the word: "jelly"
> It registers instead: "Elly"

Are you sure you're releasing J before typing E? If you press E before releasing J, the behavior is intentional. I'm not sure how you would distinguish that from J(Shift)-E.  (Would you wait for 10s before pressing E to emit a capital E? I'm guessing you wouldn't.)

If you start xremap with `RUST_LOG=debug xremap`, it will print the sequence of events that it recognized. 0 (release) of J should appear earlier than 1 (press) of E for that to work.

I admit that this feature is not documented well. If you can file a pull request to README.md that explains your current understanding, I can correct and/or merge it.

**Comment by neophrema** [comment]
I'm not sure if I release it before, cause its splits of seconds. That's why there is a need for a human-possible-timeframe, e.g. time threshold. Yet, fluent, semi fluent typing shall be possible, without erroneous capital letter input

Hypothetically, yes, I want to wait 10s before inputting capital E (realistically I would set it to 0.5 - 0.8 seconds).
Is the current state that the variable _alone_timeout_millis_ sets the time how long a key is pressed before the mapping of the key, output, gets changed? Why does the value 15000 not result in 15 sec of delay?
Is there _held_timeout_millis_?

I as soon as I understood the way it works I could do so and write a little documentation, I don't know Rust, but I might understand parts of the code.

EDIT: I'm sorry, you ask me to write down my understanding -as now- first in the Readme?

**Comment by k0kubun** [comment]
> EDIT: I'm sorry, you ask me to write down my understanding -as now- first in the Readme?

Yes, just a text explanation on README.md. Maybe replace `# Dispatch different keys depending on whether you hold it or press it alone` with something more elaborate?

**Comment by neophrema** [comment]
Hello, before I file a pull, I'd like to discuss, I propose the following, I have made little changes:

> 
> **modmap**
> 
> `modmap` is for key-to-key remapping like xmodmap. Note that remapping a key to a modifier key, e.g. CapsLock to Control_L, is supported only in `modmap` since `keymap` handles modifier keys differently.
> 
> ```
> modmap:
>   - name: Name # Optional
>     exact_match: false # Optional, defaults to: false
>     remap: # Required
>       # Replace a key with another
>       KEY_XXX: KEY_YYY # Required
>       # Dispatch a key differently depending on whether you tap or hold it. 
>       KEY_XXX:
>         tapped: KEY_ZZZ # Required
>         held: KEY_YYY # Required
>         tap_timeout_millis: 1000 # Optional, default 1000, timeframe in which a key is considered tapped.
>       # Hook `keymap` action on key press/release events.
>       KEY_XXX:
>         press: { launch: ["xdotool", "mousemove", "0", "7200"] } # Required
>         release: { launch: ["xdotool", "mousemove", "0", "0"] } # Required
>     application: # Optional
>       not: [Application, ...]
>       # or
>       only: [Application, ...]
> ```
> 
> For `KEY_XXX` etc., insert your Keyname. The keyname is case-insensitive and you can skip the `KEY_` prefix. So `KEY_CAPSLOCK`, `CAPSLOCK`, and `CapsLock` point to the same key. Some custom aliases like `SHIFT_R`, `CONTROL_L`, etc. are provided.
> 
> If you specify a map containing `held:` and `tapped:` you can use the key for two purposes. The key is considered tapped if it's pressed and released within tap_timeout_millis, _before any other key is pressed_. Otherwise it's considered held and might be used in conjunction with other keys. *
> 


1. I propose to rename the "alone" keyword because it seems misleading. What do you think? To mitigate breaking with existing configs might it be possible to make an alias, `alone_timeout_millis`=`tap_timeout_millis`?
2. I also propose, yet can't implement, another timeout: `hold_timeout_millis`.

This is directly related to my use case.

> *[You can also specify `hold_timeout_millis`: this sets the time a key has to be held before it registers as held. This might especially be useful if you want to remap letter keys, without disturbing fluent typing.]

3. If `hold_timeout_millis` gets implemented, the defaulting to alone_timeout_millis needs to be cut. Alternatively it gets overwritten if one of both variables gets set.


**EDIT:** Mini change in wording.

**Comment by k0kubun** [comment]
> I propose to rename the "alone" keyword because it seems misleading. What do you think? 

~I do agree that it's a little bit better name, but not significant enough to attempt a rename.~ (edit: see the next comment) I'd like to point out that the `held`/`alone` terminology is named after [the same thing in Karabiner-Elements](https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/to-if-alone/), which I use for macOS and means that it's somewhat accepted use of the word given how popular that is on that platform.

> *[You can also specify hold_timeout_millis: this sets the time a key has to be held before it registers as held. This might especially be useful if you want to remap letter keys, without disturbing fluent typing.]

Sounds like a good idea for your use case. I'm open to merging contributions that implement this feature. To make it easy to understand the relationship between configs, I'd like to use the same prefix here, i.e. `held_timeout_millis`, instead of deviating the name from `held`. Consistency seems more useful for configuration syntax than being grammatically perfect.

Now that we mentioned Karabiner-Elements, I looked at [its equivalent](https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/to-if-held-down/) and thought `held_threshold_millis` might be better than "timeout" for this feature.

**Comment by k0kubun** [comment]
> I do agree that it's a little bit better name, but not significant enough to attempt a rename.

I take this back. I'm open to renaming `alone` to `tapped` if you also agree with the idea about consistency; it should be `tapped_timeout_millis` instead of `tap_timeout_millis`. WDYT?

**Comment by neophrema** [comment]
Thanks for the discussion!

Yes, I was thinking about consistency,- actually it's quite funny! I wanted to propose `held_threshold_millis`, but thought I might annoy you by proposing too much change in naming.
I'm fine with either `tapped_timeout_millis`, or `tap_timeout_millis`. You decide.

I don't know Rust, have only minor knowledge of Python, my major is Philosophy. But this seems like a little programming task, if I'm not mistaken.

I'll have something big on the 20th, which I need to prepare, afterwards I'll have time to look into Rust, maybe before, but I need to be cautious since I'm fond of procrastination.
Please don't close this, I'll try to manage it.

_**If anyone else proposes a solution, feel free to submit it.**_
As soon as the Code can be merged I'll edit the readme.

@k0kubun the other minor changes I did to the excerpt are ok with you?

Thanks and till soon!

**Comment by k0kubun** [comment]
> I'm fine with either tapped_timeout_millis, or tap_timeout_millis. You decide.

:thinking: ... actually, now that we're deviating from the naming of Karabiner-Elements, instead of using a relatively long "tapped", I might want to change them to:

* `tap`
* `tap_timeout_millis`
* `hold`
* `hold_threshold_millis`

> But this seems like a little programming task, if I'm not mistaken.
> Please don't close this, I'll try to manage it.

Sure :+1:

It's okay to drop the support for old config names. Then implementing this would be probably not too difficult.

> the other minor changes I did to the excerpt are ok with you?

I'm guessing we're already on the same page. Please feel free to start working on a pull request at any time :slightly_smiling_face:

**Comment by neophrema** [comment]
Just to inform, I'm still looking into it but have some obligations, don't close please. I need some more time.
Kind regards,
v.

**Comment by stuart-warren** [comment]
I'm also trying to replicate zmk keybinds on my home row and struggling with repeatedly throwing control/alt/shift combos while typing at a normal speed

zmk seems to have flavours of these keys
* tap-preferred
* hold-preferred
* balanced
* tap-unless-interrupted

https://zmk.dev/docs/behaviors/hold-tap#flavors

**Comment by EnierGonzalezIbarra** [comment]
Any advance in this issue? I have the same problem in which typing too fast fires the "held" action when i don't want it to. alone_timeout_millis sets a timeout only when no other key is pressed

**Comment by k0kubun** [comment]
Nobody has filed a PR for it, but I'm already open to merging it once it is filed.

**Comment by hpccc53** [comment]
Tap-preferred is added in `v0.14.15`.

Does that solve the problems?
